### PR TITLE
Explicit do delegate annotation

### PR DESCRIPTION
**Pull request by sudiksha27**
[Reference Issue](https://github.com/palantir/atlasdb/issues/4411)

**Goals**:
- Fail on generating an AutoDelegate implementation if there are any default methods that have not been annotated with either @DoNotDelegate or @DoDelegate
- Do not autoDelegate CassandraKeyValueServiceConfig

**Implementation Description (bullets)**:
- Validation check is added (validateMethodsToBeAutoDelegated) in the AutoDelegateProcessor.
- It enforces decision on whether a default method should be autoDelegated or not.
- In case where a default method is not explicitly annotated, autoDelegation for the interface fails.

**Testing (What was existing testing like?  What have you done to improve it?)**:
Work in progress.

**Where should we start reviewing?**:
AutoDelegateProcessor

### CODE REVIEW

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java [hunk 1]
```diff
@@ -139,6 +141,20 @@ private static TypeElement validateAnnotatedElement(Element annotatedElement) th
         return (TypeElement) annotatedElement;
     }
 
+    private static void validateMethodsToBeAutoDelegated(Element annotatedElement, TypeToExtend typeToExtend)
+            throws ProcessingException {
+        Set<ExecutableElement> methodsToBeDelegated = typeToExtend.getMethods();
+
+        for (ExecutableElement methodElement : methodsToBeDelegated) {
+            if (methodElement.getModifiers().contains(Modifier.DEFAULT)
+                && methodElement.getAnnotation(DoDelegate.class) == null
```
Comment (Jolyon-S) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java]: I haven't checked, but is there a neater method for doing `methodElement.getAnnotation(DoDelegate.class) == null`, like maybe `checkAnnotationPresent` or something like that?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraReloadableKvsConfig.java [hunk 2]
```diff
@@ -62,12 +157,102 @@ public int fetchBatchCount() {
...
+    @Override
+    public boolean usingSsl() {
+        return config.usingSsl();
+    }
+
+    @Override
+    public void check() {
+
+    }
+
```
Comment (Jolyon-S) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraReloadableKvsConfig.java]: this feels like a lot of copy and pasted methods, but I'm not sure if there is a better way of doing this while avoiding delegates.

Comment (OStevan) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraReloadableKvsConfig.java]: I don't really think we need this here. The worst thing that can happen is that we call the check as defined in the `CassandraKeyValueServiceConfig` which also in a way makes sense.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraReloadableKvsConfig.java [hunk 3]
```diff
@@ -62,12 +157,102 @@ public int fetchBatchCount() {
...
+    @Override
+    public boolean usingSsl() {
+        return config.usingSsl();
+    }
+
+    @Override
+    public void check() {
+
```
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/CassandraReloadableKvsConfig.java]: nit: Might be good to leave a comment explaining why `check()` is empty. _In this case, this is correct._

File: atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java [hunk 4]
```diff
@@ -139,6 +142,19 @@ private static TypeElement validateAnnotatedElement(Element annotatedElement) th
...
     }
 
+    private static void validateMethodsToBeAutoDelegated(Element annotatedElement, TypeToExtend typeToExtend)
+            throws ProcessingException {
+        Set<ExecutableElement> methodsToBeDelegated = typeToExtend.getMethods();
+
+        for (ExecutableElement methodElement : methodsToBeDelegated) {
+            if (methodElement.getModifiers().contains(Modifier.DEFAULT)
+                    && methodElement.getAnnotation(DoDelegate.class) == null) {
+                throw new ProcessingException(annotatedElement, "Default methods must be annotated with "
```
Comment (jeremyk-91) [inline atlasdb-processors/src/main/java/com/palantir/processors/AutoDelegateProcessor.java]: Would be good to explain why we don't need to check for `DoNotDelegate` here.

**Pull request merged**